lib/src/authentication/credential_store.dart

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import 'dart:convert';
+
+import 'package:path/path.dart' as path;
+
+import '../io.dart';
+import '../system_cache.dart';
+import '../utils.dart';
+import 'credential.dart';
+
+class CredentialStore {
+  CredentialStore(this.cache);
+
+  final SystemCache cache;
+
+  /// Adds [credentials] for [server] into store.
+  void addServer(String server, Credential credentials) {
+    var key = server.toLowerCase();
+    // Make sure server name ends with a backslash. It's here to deny possible
+    // credential thief attach vectors where victim can add credential for
+    // server 'https://safesite.com' and attacker could steal credentials by
+    // requesting credentials for 'https://safesite.com.attacker.com', because
+    // URL matcher (_serverMatches method) matches credential keys with the
+    // beginning of the URL.
+    if (!key.endsWith('/')) key += '/';
+    serverCredentials[key] = credentials;
+    _save();
+  }
+
+  /// Removes credentials for servers that [url] matches with.
+  void removeServer(String url) {
+    serverCredentials.removeWhere((key, value) => serverKeyMatches(key, url));
+    _save();
+  }
+
+  /// Returns pair of credential and server key for server that [url] and
+  /// [alsoMatches] matches to server key.
+  Pair<String, Credential>? getCredential(String url,
+      {List<String>? alsoMatches}) {
+    for (final key in serverCredentials.keys) {
```

Comment:
I think words like `key` and `serverKey` are unfortunate here.. might be best to call it `serverUrl`, `serverBaseUrl` or something like that.

In a "credential store" I associate "key" with a secret private/public key. In this case it's more like _server identifier_.